plugins/out_s3/s3_store.c

```diff
@@ -179,13 +183,60 @@ int s3_store_buffer_put(struct flb_s3 *ctx, struct s3_file *s3_file,
 
         /* Use fstore opaque 'data' reference to keep our context */
         fsf->data = s3_file;
+        created = FLB_TRUE;
     }
     else {
         fsf = s3_file->fsf;
     }
 
-    /* Append data to the target file */
-    ret = flb_fstore_file_append(fsf, data, bytes);
+    /*
+     * Append data to the target file
+     *
+     * If the user desires CSV format, we do the formatting at this level since before this
+     * step we don't know when we will be appending to the beginning of a new file so we can
+     * add the CSV headers.
+     */
+    if (ctx->format == S3_FORMAT_CSV) {
+        if (created && ctx->csv_column_names) {
+            csv_add_columns = FLB_TRUE;
+        }
+
+        if (ctx->log_key) {
+            /*
+             * if log_key option was used, the incoming data is NOT a msgpack object,
+             * likely a JSON string message that we need to convert back so we can
+             * do the CSV formatting.
```

Comment:
I'm struggling to understand how CSV works with log_key.

So the incoming msgpack is an array of records, each which might have multiple keys. I thought the goal of the CSV column headers is to give names to each of the top level keys in each record? With log key, we just take a single key from each record called `log`. 

In the log_key case, the incoming data is not a JSON. Its a newline delimited buffer of each record's log_key value. That is the value of each log key for each record delimitted by newlines. You can't parse it as JSON unless you do it line by line. Each line might be a JSON or it might just be a string.

Reply:
As far as I understand it, when `log` key is set it contains the full log in plain text. If that is so we would need to parse it with a fluent-bit parser to transform it into a set of fields.

Perhaps one or several of the following might be a better solution?

- Add a new `log_key_parser` option to use a parser for the log_key.
- Add a new `log_key_format` option to parse it as one of several specific formats.
- Send the `log_key` directly as a line of CSV.